src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSubject.cs

```diff
@@ -29,148 +29,209 @@
 using System.Collections.ObjectModel;
 using System.Security.Claims;
 using Microsoft.IdentityModel.Logging;
+using static Microsoft.IdentityModel.Logging.LogHelper;
+using Microsoft.IdentityModel.Xml;
 
 namespace Microsoft.IdentityModel.Tokens.Saml
 {
+    /// <summary>
+    /// Represents the Subject element specified in [Saml2Core, 2.4.2.1].
+    /// </summary>
+    /// <remarks>
+    /// If the NameId is null and the SubjectConfirmations collection is empty,
+    /// an InvalidOperationException will be thrown during serialization.
+    /// </remarks>
     public class SamlSubject
     {
         // Saml SubjectConfirmation parts.
-        Collection<string> _confirmationMethods = new Collection<string>();
         string _confirmationData;
-        SecurityKeyIdentifier _securityKeyIdentifier;
         SecurityKey _crypto;
+        private KeyInfo _keyInfo;
 
         // Saml NameIdentifier element parts.
         string _name;
         string _nameFormat;
         string _nameQualifier;
 
         // TODO remove this internal
-        internal SamlSubject()
+        /// <summary>
+        /// Initialize an instance of <see cref="SamlSubject"/>.
+        /// </summary>
+        public SamlSubject()
         {
         }
 
+        /// <summary>
+        /// Initialize an instance of <see cref="SamlSubject"/>.
+        /// </summary>
+        /// <param name="nameFormat">The format of the subject.</param>
+        /// <param name="nameQualifier">The NameIdentifier of the subject.</param>
+        /// <param name="name">The name of the subject.</param>
         public SamlSubject(string nameFormat, string nameQualifier, string name)
-            : this(nameFormat, nameQualifier, name, null, null, null)
+            : this(nameFormat, nameQualifier, name, null, null/*, null*/)
         {
         }
 
+        /// <summary>
+        /// Initialize an instance of <see cref="SamlSubject"/>.
+        /// </summary>
+        /// <param name="nameFormat">The format of the subject.</param>
+        /// <param name="nameQualifier">The NameIdentifier of the subject.</param>
+        /// <param name="name">The name of the subject.</param>
+        /// <param name="confirmations">The confirmation methods contained in the subject.</param>
+        /// <param name="confirmationData">The confirmation data contained in the subject.</param>
         public SamlSubject(
             string nameFormat,
             string nameQualifier,
             string name,
             IEnumerable<string> confirmations,
-            string confirmationData,
-            SecurityKeyIdentifier securityKeyIdentifier)
+            string confirmationData)
+            //SecurityKeyIdentifier securityKeyIdentifier)
         {
             if (confirmations != null)
             {
```

Comment:
Yeah, the new List(Ienumerable) has better performance but cannot loop through and check all the values.